google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITBlobWriteChannelTest.java

```diff
@@ -193,4 +198,113 @@ private void doJsonUnexpectedEOFTest(int contentSize, int cappedByteCount) throw
     ByteBuffer actual = ByteBuffer.wrap(actualData.toByteArray());
     assertEquals(expected, actual);
   }
+
+  @Test
+  public void blobWriteChannel_handlesRecoveryOnLastChunkWhenGenerationIsPresent_multipleChunks()
+      throws IOException {
+    int _2MiB = 256 * 1024;
+    int contentSize = 292_617;
+
+    blobWriteChannel_handlesRecoveryOnLastChunkWhenGenerationIsPresent(_2MiB, contentSize);
+  }
+
+  @Test
+  public void blobWriteChannel_handlesRecoveryOnLastChunkWhenGenerationIsPresent_singleChunk()
+      throws IOException {
+    int _4MiB = 256 * 1024 * 2;
+    int contentSize = 292_617;
+
+    blobWriteChannel_handlesRecoveryOnLastChunkWhenGenerationIsPresent(_4MiB, contentSize);
+  }
+
+  private void blobWriteChannel_handlesRecoveryOnLastChunkWhenGenerationIsPresent(
+      int chunkSize, int contentSize) throws IOException {
+    Instant now = Clock.systemUTC().instant();
+    DateTimeFormatter formatter =
+        DateTimeFormatter.ISO_LOCAL_DATE_TIME.withZone(ZoneId.from(ZoneOffset.UTC));
+    String nowString = formatter.format(now);
+    BucketInfo bucketInfo = BucketInfo.of(dataGeneration.getBucketName());
+    String blobPath = String.format("%s/%s/blob", testName.getMethodName(), nowString);
+    BlobId blobId = BlobId.of(dataGeneration.getBucketName(), blobPath);
+    BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+
+    ByteBuffer contentGen1 = dataGeneration.randByteBuffer(contentSize);
+    ByteBuffer contentGen2 = dataGeneration.randByteBuffer(contentSize);
+    ByteBuffer contentGen2Expected = contentGen2.duplicate();
+    Storage storage = StorageOptions.getDefaultInstance().getService();
```

Comment:
In general I think it makes sense to move these tests over to this class. However, I do want to avoid mixing testbench and live service tests in the same class (unless there is something external controlling the storage instance).

In this case, we can refactor this test slightly to have it run against the testbench, and still validate what it is meant to. This suggestion, plus the two following should all be applied together.

```suggestion
    HttpStorageOptions baseStorageOptions =
        StorageOptions.http()
            .setCredentials(NoCredentials.getInstance())
            .setHost(testBench.getBaseUri())
            .setProjectId("test-project-id")
            .build();
    Storage storage = baseStorageOptions.getService();
```